src/E1_coldflow/E1_coldflow.cpp

```diff
@@ -105,26 +100,25 @@ void loop() {
     int i = 0;
     while (RFSerial.available()) {
       command[i] = RFSerial.read();
-      Serial.print(command[i]);
       i++;
     }
+    Serial.write(command, i);
 
     debug(String(command), DEBUG);
-    int action = decode_received_packet(String(command), &valve, valves, numValves);
+    int action = decode_received_packet(command, &valve, valves, numValves, DEBUG);
     if (action != -1) {
       take_action(&valve, action);
-      packet = make_packet(valve.id, false);
-      Serial.println(packet);
+      packet_size = make_packet(valve.id, false);
+      Serial.write(packet, packet_size);
       #if SERIAL_INPUT != 1
-        RFSerial.println(packet);
+        RFSerial.write(packet, packet_size);
       #endif
-      write_to_SD(packet.c_str(), file_name);
+      packet[packet_size] = '\0';
```

Comment:
What is this for? Is this a delimiter?

Reply:
If so, why do we need it? We already have the packet start and packet end bytes

Reply:
`write_to_sd` expects input of the format c String which is a character array terminated by the null character `\0`.

Reply:
oh I see. `\0` is the byte `0x00` tho, so what happens if the packet has a zero byte in it?

Reply:
fixed. It takes a length now